File: service/javadsl/kinesis/client/src/main/java/com/lightbend/lagom/javadsl/broker/kinesis/KinesisTopicFactory.java
/*
 * Copyright (C) 2017 Quantify Labs Inc. <https://github.com/streetcontxt>
 */
package com.lightbend.lagom.javadsl.broker.kinesis;

import javax.inject.Inject;

import com.lightbend.lagom.internal.broker.kinesis.KinesisConfig;
import com.lightbend.lagom.internal.broker.kinesis.KinesisConfig$;
import com.lightbend.lagom.internal.javadsl.api.broker.TopicFactory;
import com.lightbend.lagom.internal.javadsl.broker.kinesis.JavadslKinesisTopic;
import com.lightbend.lagom.javadsl.api.Descriptor;
import com.lightbend.lagom.javadsl.api.ServiceInfo;
import com.lightbend.lagom.javadsl.api.ServiceLocator;
import com.lightbend.lagom.javadsl.api.broker.Topic;

import akka.actor.ActorSystem;
import akka.stream.ActorMaterializer;
import scala.concurrent.ExecutionContext;

/**
 * Factory for creating topics instances.
 */
public class KinesisTopicFactory implements TopicFactory {
    private final ServiceInfo serviceInfo;
    private final ActorSystem system;
    private final ActorMaterializer materializer;
    private final ExecutionContext executionContext;
    private final KinesisConfig config;
    private final ServiceLocator serviceLocator;

    @Inject
    public KinesisTopicFactory(ServiceInfo serviceInfo, ActorSystem system, ActorMaterializer materializer,
            ExecutionContext executionContext, ServiceLocator serviceLocator) {
        this.serviceInfo = serviceInfo;
        this.system = system;
        this.materializer = materializer;
        this.executionContext = executionContext;
        this.config = KinesisConfig$.MODULE$.apply(system.settings().config());
        this.serviceLocator = serviceLocator;
    }

    @Override
    public <Message> Topic<Message> create(Descriptor.TopicCall<Message> topicCall) {
        return new JavadslKinesisTopic<>(config, topicCall, serviceInfo, system, serviceLocator, materializer, executionContext);
    }
}

File: service/scaladsl/kinesis/client/src/main/scala/com/lightbend/lagom/internal/scaladsl/broker/kinesis/KinesisTopicFactory.scala
/*
 * Copyright (C) 2017 Quantify Labs Inc. <https://github.com/streetcontxt>
 */
package com.lightbend.lagom.internal.scaladsl.broker.kinesis

import akka.actor.ActorSystem
import akka.stream.Materializer
import com.lightbend.lagom.internal.broker.kinesis.KinesisConfig
import com.lightbend.lagom.internal.scaladsl.api.broker.TopicFactory
import com.lightbend.lagom.scaladsl.api.Descriptor.TopicCall
import com.lightbend.lagom.scaladsl.api.broker.Topic
import com.lightbend.lagom.scaladsl.api.{ServiceInfo, ServiceLocator}

import scala.concurrent.ExecutionContext

/**
 * Factory for creating topics instances.
 */
private[lagom] class KinesisTopicFactory(serviceInfo: ServiceInfo, system: ActorSystem,
  serviceLocator: ServiceLocator)(implicit
  materializer: Materializer,
  executionContext: ExecutionContext) extends TopicFactory {

  private val config = KinesisConfig(system.settings.config)

  def create[Message](topicCall: TopicCall[Message]): Topic[Message] = {
    new ScaladslKinesisTopic(config, topicCall, serviceInfo, system, serviceLocator)
  }
}
